### PR TITLE
100 create a question creation view

### PR DESCRIPTION
**Pull request by jaess105**
# Task

Create a view in which questions can be created.  
Currently only supports:
* free-text
* single-choice
* multiple-choice

## Functionality
* GET
  Displays a form for creating a question, in which previous created categories assigned to other questions are displayed as well.
Three different types of questions can be created with this form. One can switch between them using a drop-down. The contents are always present, but via JS they are displayed or not. On post, the hidden values are removed from the post request to only send one type of question.
New Categories can be added as well, and assigned to the newly created question.
* POST
Tries to parse the data, then validates that data and if everything worked, saves the question in the db and displayes a success view.
If there were parsing or validation errors, the question-creation-form is displayed again with the errors, and the data that was previously sent, displayed in the form as well.
(This controller does not redirect to the get request again, as it would lead to a lot of data send via the parameters which is not suitable in my opinion.)

## Changes
* Updated Model
  * Added categories spec to question spec.
  * Updated question specifications that weren't specific enough
* Added create-question-view
  * added create-question-form for creating a question
  * added create-question-success for displaying successfully created questions
* Added QuestionService and QuestionService protocol
* Updated db
  * Added function `get-all-question-categories`  to fetch the newly added question/categories from the database.
  * Added categories to dummy data
* Added question-controller
* Core
  * Added get and post routes to core
  * Added QuestionService instance
* Added Tests for:
  * create-question-view
  * question-service
  * question-controller

## Other changes
* Included js into resources/public:
  * Added folder to Ring, to make them available in htmls
  * Added folder to Docker, to make it available when building
  * Removed folder from .gitignore
* Moved utility functions that are used by other controllers from the course-iteration-controller to corresponding files.
  * Created hiccup_extensions
  * Enhanced ring_extensions
    * Included bootstrap into html-response

### CODE REVIEW

**Review by MarkusBott — COMMENTED** [review]
looks fine, can be merged if you and phillipp discussed the Javascript part

File: src/clj/services/question_service/p_question_service.clj [hunk 1]
```diff
@@ -0,0 +1,23 @@
+(ns services.question-service.p-question-service)
+
+
+(defprotocol PQuestionService
+
+  (create-question!
+    [self question]
+    "Creates a course-iteration entry in the database and returns a course-iteration map.")
```
Comment (hehin101) [inline src/clj/services/question_service/p_question_service.clj:8]: I think it should be a "question entry".

File: src/clj/util/hiccup_extensions.clj [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    [hiccup2.core :as h]))
+
+
+(s/fdef script
+        :args (s/cat :script-src (s/* string?))
+        :ret  #(instance? clojure.lang.PersistentVector %))
+
+
+(defn script
+  "Takes any number of and joins them into a javascript script tag.
```
Comment (hehin101) [inline src/clj/util/hiccup_extensions.clj:14]: Any number of what?

**Comment by jaess105** [comment]
Philipp and I discussed it and we came to the conclusion that it is ok to leave the JS part in for now. But in the long run we want to replace it by cljs. 
I added some comments stating this with the last commit and opened an [Issue](https://github.com/pkoerner/expert-parakeet/issues/120) for the replacement.

**Comment by jaess105** [comment]
The `db/conn?` issue was related to this [issue]( https://github.com/pkoerner/expert-parakeet/issues/112). I'll merge the current main into this branch again. This should fix it.

**Comment by jaess105** [comment]
The db/conn? issue is fixed

File: test/controllers/question/question_controller_test.clj [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+        {(str :question/type) "No valid type"}
+        {(str :question/question-statement) "No valid question statement!"}
+        {(str :question/points) "No valid question points"}
+        {(str :question/categories) "No valid question categories"}
+        {(str :question/evaluation-criteria) "No valid evaluation criteria"}
+        {(str :question/possible-solutions) "No valid possible solutions"}
+        {(str :question/single-choice-solution) "No valid single choice solution"}
+        {(str :question/multiple-choice-solution) "No valid multiple choice solutions"}
+        {(str :question/multiple-choice-solution) "No valid multiple choice solutions"}))
```
[the remark this answers is not in the document]
Comment (jaess105) [reply test/controllers/question/question_controller_test.clj]: Yes, we need to be extra certain here. Kidding, removed the duplicated check.

File: test/controllers/question/question_controller_test.clj [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+                     question-service-stub (stub-question-service
+                                             :create-question! db-add-fun-stub
+                                             :validate-question (partial validate-question question-service))]
+                 (submit-create-question! (->> question
+                                               (map (fn [[key val]] {(str key) val}))
+                                               (into {})
+                                               (assoc test-request :multipart-params))
+                                          redirect-uri
+                                          question-service-stub)
+                 @was-not-called)
```
[the remark this answers is not in the document]
Comment (jaess105) [reply test/controllers/question/question_controller_test.clj]: You're absolutely right. I changed the value the function returns to `false`

File: src/clj/services/question_service/p_question_service.clj (hunk 1 above)
Comment (jaess105) [reply src/clj/services/question_service/p_question_service.clj]: Yes, fixed the copy :spaghetti: mistake and replaced both occurrences of `course-iteration`  with `question`

File: src/clj/util/hiccup_extensions.clj (hunk 2 above)
Comment (jaess105) [reply src/clj/util/hiccup_extensions.clj]: Updated the comment. It now reads:
`"Takes any number of arguments and joins them into a string, which is returned in a javascript script tag"`

**Review by hehin101 — APPROVED** [review]

**Pull request merged**